salt/modules/virt.py

```diff
@@ -580,90 +700,34 @@ def init(name,
-        # TODO: make it possible to use disk profiles and use the
-        # template image as the system disk
-        diskp = _disk_profile('default', hypervisor, **kwargs)
-        log.debug('Disk profile is {0}'.format(diskp))
-
-        # When using a disk profile extract the sole dict key of the first
-        # array element as the filename for disk
+    diskp = _disk_profile(disk, hypervisor, **kwargs)
+
+    if image:
+        # If image is specified in module arguments, then it will be used
+        # for the first disk instead of the image from the disk profile
         disk_name = next(six.iterkeys(diskp[0]))
-        disk_type = diskp[0][disk_name]['format']
-        disk_file_name = '{0}.{1}'.format(disk_name, disk_type)
-
-        if hypervisor in ['esxi', 'vmware']:
-            # TODO: we should be copying the image file onto the ESX host
-            raise SaltInvocationError(
-                'virt.init does not support image template template in '
-                'conjunction with esxi hypervisor'
-            )
-        elif hypervisor in ['qemu', 'kvm']:
-            img_dir = __salt__['config.option']('virt.images')
-            log.debug('Image directory from config option `virt.images` is {0}'
-                      .format(img_dir))
-            img_dest = os.path.join(
-                img_dir,
-                name,
-                disk_file_name
-            )
-            log.debug('Image destination will be {0}'.format(img_dest))
-            img_dir = os.path.dirname(img_dest)
-            log.debug('Image destination directory is  {0}'.format(img_dir))
-            sfn = __salt__['cp.cache_file'](image, saltenv)
-
-            try:
-                os.makedirs(img_dir)
-            except OSError:
-                pass
-
-            qcow2 = False
-            if salt.utils.which('qemu-img'):
-                res = __salt__['cmd.run']('qemu-img info {}'.format(sfn))
-                imageinfo = yaml.load(res)
-                qcow2 = imageinfo['file format'] == 'qcow2'
-
-            try:
-                if enable_qcow and qcow2:
-                    log.info('Cloning qcow2 image {} using copy on write'
-                              .format(sfn))
-                    __salt__['cmd.run'](
-                        'qemu-img create -f qcow2 -o backing_file={} {}'
-                        .format(sfn, img_dest).split())
+        log.debug('{0} image from module arguments will be used for disk "{1}"'
+                  ' instead of {2}'
+                  .format(image,
+                          disk_name,
+                          diskp[0][disk_name].get('image', None)))
+        diskp[0][disk_name]['image'] = image
+
+    # Create multiple disks, empty or from specified images.
+    for disk in diskp:
+        log.debug("Creating disk for VM [ {0} ]: {1}".format(name, disk))
+
+        for disk_name, args in six.iteritems(disk):
+
+            if hypervisor in ['esxi', 'vmware']:
```

Comment:
I would prefer to make such changes in a different patch, because it could introduce regression that is not connected to the feature I'm trying to add :)